File: ios/Releaf/Data/Task/TaskEntity.swift
/*
 * TaskEntity.swift
 *
 * GRDB record for the `tasks` table. Mirror of Android's
 * `TaskEntity.kt` — snake_case column names match the shared SQL
 * schema (design-system/migrations/v1_initial.sql §5) so Drive sync
 * payloads round-trip byte-for-byte between platforms.
 *
 * Named `TaskRecord` rather than `Task` to avoid colliding with the
 * Swift concurrency `_Concurrency.Task` type. Every feature-layer
 * reference should say `TaskRecord`.
 */

import Foundation
import GRDB

public struct TaskRecord: Codable, FetchableRecord, PersistableRecord,
                          Identifiable, Equatable, Sendable {

    public static let databaseTableName = "tasks"

    // MARK: - Columns

    /// UUIDv7, sorted chronologically by construction.
    public var id: String

    /// Scoping column; all queries filter by the signed-in user.
    public var userId: String

    public var title: String

    /// Optional longer notes body.
    public var description: String?

    /// Local `YYYY-MM-DD` date the task is due. The SQL CHECK
    /// constraint gates the GLOB shape — keep that in mind when
    /// synthesising values outside of DatePicker.
    public var dueDate: String?

    /// GRDB stores Bool as INTEGER 0/1, matching the SQL shape.
    public var completed: Bool

    /// ISO-8601 UTC with ms; set at the moment the user ticks the box.
    public var completedAt: String?

    /// 0 = none, 1 = low, 2 = medium, 3 = high. Mirrors
    /// `NotepadTodo.priority` so a promoted todo keeps its level.
    public var priority: Int

    /// ISO-8601 UTC with ms. See IsoClock.
    public var createdAt: String

    public var updatedAt: String

    /// `true` = needs upload to Drive; cleared by the sync worker.
    public var dirty: Bool

    /// ISO-8601 UTC when soft-deleted; nil = active.
    public var deletedAt: String?

    // MARK: - Column mapping

    public enum CodingKeys: String, CodingKey {
        case id
        case userId       = "user_id"
        case title
        case description
        case dueDate      = "due_date"
        case completed
        case completedAt  = "completed_at"
        case priority
        case createdAt    = "created_at"
        case updatedAt    = "updated_at"
        case dirty
        case deletedAt    = "deleted_at"
    }

    // MARK: - Init

    public init(
        id: String,
        userId: String,
        title: String,
        description: String? = nil,
        dueDate: String? = nil,
        completed: Bool = false,
        completedAt: String? = nil,
        priority: Int = 0,
        createdAt: String,
        updatedAt: String,
        dirty: Bool = true,
        deletedAt: String? = nil
    ) {
        self.id = id
        self.userId = userId
        self.title = title
        self.description = description
        self.dueDate = dueDate
        self.completed = completed
        self.completedAt = completedAt
        self.priority = priority
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.dirty = dirty
        self.deletedAt = deletedAt
    }
}

File: apps/releaf/android/app/src/main/java/app/releaf/mobile/data/task/TaskEntity.kt
/*
 * TaskEntity.kt
 *
 * Room entity mirroring the `tasks` table in
 * design-system/migrations/v1_initial.sql (§5). Only the columns the
 * v1 UI actually reads are materialised here — project_id, position,
 * and the subtask/reminder tables are parked until the features that
 * need them land.
 *
 * Tasks are a workspace-level concept (distinct from the inline todos
 * that live on a page or notepad entry). A todo is "promoted" to a
 * task when it needs tracking outside its originating page; the
 * promote path will copy the todo text into a new `tasks` row and
 * keep the two in sync. For now the Tasks screen only supports
 * directly-created tasks — the promote path is tracked as follow-up.
 */

package app.releaf.mobile.data.task

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

/**
 * Kanban-style lifecycle marker, distinct from [TaskEntity.completed].
 *
 *   • [Todo]  — captured, not started. Default for new tasks.
 *   • [Doing] — actively in progress. Surfaces in the "Doing" column
 *               of the Boards view and as a visible badge on the row.
 *   • [Done]  — completed. Kept in sync with [TaskEntity.completed]
 *               (and [TaskEntity.completedAt]) at the repository
 *               layer so existing "open count" / "completed filter"
 *               queries keep working unchanged.
 *
 * Stored as the lowercase [wire] string — the default column value
 * in the schema is `'todo'`, so pre-v7 rows that don't have the
 * column migrated get materialised as [Todo] on next read.
 */
enum class TaskStatus(val wire: String) {
    Todo("todo"),
    Doing("doing"),
    Done("done");

    companion object {
        /** Lenient parse — anything we don't recognise falls back to [Todo]. */
        fun parse(raw: String?): TaskStatus = when (raw) {
            "doing" -> Doing
            "done"  -> Done
            else    -> Todo
        }
    }
}

@Entity(
    tableName = "tasks",
    indices = [
        Index("user_id"),
        Index("due_date"),
        Index("completed"),
        Index("status"),
        Index("updated_at"),
        Index("deleted_at"),
    ],
)
data class TaskEntity(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    /** Scoping column — tasks are per-user, same as notepad entries. */
    @ColumnInfo(name = "user_id")
    val userId: String,

    @ColumnInfo(name = "title")
    val title: String,

    /** Longer free-form notes. Optional. */
    @ColumnInfo(name = "description")
    val description: String? = null,

    /**
     * Local YYYY-MM-DD the task is due on. Null = no due date. The
     * schema CHECK constraint gates the GLOB shape; Room doesn't
     * reproduce that here but every write path in this module goes
     * through IsoClock / format helpers that preserve it.
     */
    @ColumnInfo(name = "due_date")
    val dueDate: String? = null,

    /** Room stores Bool as INTEGER 0/1, matching the SQL shape. */
    @ColumnInfo(name = "completed", defaultValue = "0")
    val completed: Boolean = false,

    /** ISO-8601 UTC with ms; set at the moment the user ticks the box. */
    @ColumnInfo(name = "completed_at")
    val completedAt: String? = null,

    /**
     * 0 = none, 1 = low, 2 = medium, 3 = high — mirrors
     * `TodoItem.priority` so a promoted todo keeps its level.
     */
    @ColumnInfo(name = "priority", defaultValue = "0")
    val priority: Int = 0,

    /**
     * Kanban lifecycle marker — see [TaskStatus]. Stored as its wire
     * string (`"todo"` / `"doing"` / `"done"`). Kept in sync with
     * [completed] at the repository layer — setting status=done also
     * flips [completed] + [completedAt] so the existing sort / count
     * queries still work without knowing about status.
     */
    @ColumnInfo(name = "status", defaultValue = "todo")
    val status: String = TaskStatus.Todo.wire,

    /** ISO-8601 UTC with ms. See IsoClock. */
    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String,

    /** 1 = needs upload to Drive; cleared by sync worker. */
    @ColumnInfo(name = "dirty", defaultValue = "1")
    val dirty: Boolean = true,

    /** ISO-8601 UTC when soft-deleted; null = active. */
    @ColumnInfo(name = "deleted_at")
    val deletedAt: String? = null,
)
